frontend/src/queries/Query/Query.tsx

```diff
@@ -0,0 +1,40 @@
+import { isDataNode, isEventsTableNode, isLegacyQuery } from '../utils'
+import { LegacyInsightQuery } from '~/queries/nodes/LegacyInsightQuery'
+import { EventsTableQuery } from '~/queries/nodes/EventsTableQuery'
+import { DataNodeQuery } from '~/queries/nodes/DataNodeQuery'
+import { Node } from '~/queries/schema'
+import { ErrorBoundary } from '~/layout/ErrorBoundary'
+
+export interface QueryProps {
+    query: Node | string
+    setQuery?: (node: Node) => void
+}
+
+export function Query(props: QueryProps): JSX.Element {
+    const { query, setQuery } = props
+    if (typeof query === 'string') {
+        try {
+            return <Query {...props} query={JSON.parse(query)} />
+        } catch (e: any) {
+            return <div className="border border-danger p-4 text-danger">Error parsing JSON: {e.message}</div>
+        }
+    }
+    let component
+    if (isLegacyQuery(query)) {
```

Comment:
Annoying bit with if-else is that it's hard to ensure that the matching is exclusive. Would it make sense to switch this to a `switch` statement?

Reply:
Nope, as switch statements don't scope down types from my experience :(